cpp/src/copying/copy.cu

```diff
@@ -114,6 +143,125 @@ struct copy_if_else_functor_impl<string_view> {
+ * if either `lhs` or `rhs` is a scalar, partially because:
+ *   1. Struct scalars are not yet available.
+ *   2. List scalars do not yet support explosion to a full column.
+ *
+ * TODO: Using `if constexpr` to collapse this with its other overload
+ *       causes `missing return statement` warnings. Find way around this.
+ */
+template <typename Left,
+          typename Right,
+          typename Filter,
+          CUDF_ENABLE_IF(std::is_same<Left, cudf::scalar>::value ||
+                         std::is_same<Right, cudf::scalar>::value)>
+std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
+                                                     Right const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FAIL("Scalars of nested types are not currently supported!");
+}
+
+/**
+ * @brief Functor to generate gather-map for LHS column
+ *
+ * If specified `Predicate` evaluates to `true` for index `i`,
+ * gather map must contain `i` (to select LHS[i]).
+ * If false, gather map must have `null_index`, so that a null
+ * is gathered in its place.
+ */
+template <typename Predicate>
+class lhs_gather_map_functor {
+ public:
+  lhs_gather_map_functor(Predicate predicate, size_type null_index)
+    : _pred(predicate), _null_index(null_index)
+  {
+  }
+
+  size_type __device__ operator()(size_type i) const { return _pred(i) ? i : _null_index; }
+
+ private:
+  Predicate _pred;
+  size_type _null_index;
+};
+
+/**
+ * @brief Adapter to negate predicates.
+ */
+template <typename Predicate>
+class logical_not {
+ public:
+  explicit logical_not(Predicate predicate) : _pred{predicate} {}
+
+  bool __device__ operator()(size_type i) const { return not _pred(i); }
+
+ private:
+  Predicate _pred;
+};
+
+/**
+ * @brief Implementation of copy_if_else() with gather()/scatter().
+ *
+ * Currently supports only nested-type column_views. Scalars are not supported.
+ */
+template <typename Left,
+          typename Right,
+          typename Filter,
+          CUDF_ENABLE_IF(not std::is_same<Left, cudf::scalar>::value &&
+                         not std::is_same<Right, cudf::scalar>::value)>
+std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
+                                                     Right const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto null = size + 1;  // Out of bounds index, for gather() to nullify.
```

Comment:
```suggestion
  auto const null_map_entry = size + 1;  // Out of bounds index, for gather() to nullify.
```

I'm nervous about overloading the name null in general.

Reply:
Fixed.

Reply:
Yikes. I'd forgotten the `const` when I resolved this. @nvdbaranec caught it also.